src/main/java/io/cdap/plugin/scd2/SCD2Iterator.java

```diff
@@ -60,22 +62,64 @@ protected StructuredRecord computeNext() {
     cur = next != null ? next : records.next();
     next = records.hasNext() ? records.next() : null;
 
+    deduplicateCurRecord();
+    deduplicateNextRecord();
+
+    // if key changes, clean up the table to free memory
+    if (prev != null && !prev._1().equals(cur._1())) {
+      valTable.row(prev._1()).clear();
+    }
+
+    return computeRecord(cur._1(),
+                         prev != null && prev._1().equals(cur._1()) ? prev._2() : null,
+                         cur._2(),
+                         next != null && next._1().equals(cur._1()) ? next._2() : null);
+  }
+
+  /**
+   * Find the cur record that will not get deduplicated. This record will be used to returned and the end time
+   * will be computed for it.
+   *
+   * For example:
+   * 1. id: 1, name: john, isTarget: false (should get dedup)
+   * 2. id: 1, name: john, isTarget: true (should be preserved)
+   * 3. id: 1, name: john, isTarget: false (should get dedup)
+   * 4. id: 1, name: sam, isTarget: false (should get dedup)
+   * 5. id: 1, name: sam, isTarget: false (should be preserved)
+   *
+   * After this method, the prev, cur, next will be pointed to:
+   * 1.(prev) id: 1, name: john, isTarget: false (should get dedup)
+   * 2.(cur) id: 1, name: john, isTarget: true (should be preserved and returned)
+   * 3.(next) id: 1, name: john, isTarget: false (should get dedup)
+   * 4. id: 1, name: sam, isTarget: false (should get dedup)
+   * 5. id: 1, name: sam, isTarget: false (should be preserved)
+   */
+  private void deduplicateCurRecord() {
     boolean isDiff = false;
     // deduplicate the result
     while (conf.deduplicate() && !isDiff) {
-      if (next == null || !next._1().equals(cur._1())) {
+      // if this record is coming from scd2 target table, do not dedup
+      StructuredRecord curRecord = cur._2();
+      if (conf.preserveTarget() && (Boolean) curRecord.get(conf.getIsTargetField())) {
+        break;
+      }
+
+      SCD2Key curKey = cur._1();
+      SCD2Key nextKey = next._1();
+      if (next == null || !nextKey.equals(curKey)) {
         break;
       }
 
-      for (Schema.Field field : cur._2().getSchema().getFields()) {
+      for (Schema.Field field : curRecord.getSchema().getFields()) {
```

Comment:
refactor this into a separate recordsEqual() method for checking record equality, ignoring fields in the blacklist